skytap/course-manager-examples
# Metadata Stub Server - Realistic behavior (2594)

https://skytap.atlassian.net/browse/CMGR-2594

## Acceptance Criteria

- Metadata Stub Server behaves more like CM control endpoint for metadata management
  - Able to Update / Create metadata (Merges with existing metadata)
  - Able to Delete metadata (pass null for the value)

### How to Test

Not sure how you test the stub server, but what I did was:
- Add `skytap-metadata` to 	`etc/hosts` pointing to `127.0.0.1`
- Booted up sinatra locally
- Use postman to send PUT requests to `http://skytap-metadata/lab_access/self_learner/1/111/control/1/111`
- First add some metadata to various levels:
```JSON
{
    "metadata": { "AcmeDataProUsername": "user_assigned_from_script" },
    "sensitive_metadata": { "AcmeDataProPassword": "password_assigned_from_script" },
    "course": {
	    "metadata":  { "Latest Lab Provision": "script_identifier" },
	    "sensitive_metadata":  { "Other variable": "bar" }
    },
    "user": {
	    "metadata":  { "Latest user Lab Provision": "script_identifier" },
	    "sensitive_metadata":  { "Other user variable": "bar" }
    }
}
```
- Second try adding / updating / deleting metadata to various levels:
```JSON
{
    "metadata": { "AcmeDataBasicUsername": "basic_user_assigned_from_script" },
    "sensitive_metadata": { "AcmeDataProPassword": "password_updated_from_script" },
    "course": {
   		"sensitive_metadata":  { "Secret variable": "foo" }
    },
    "user": {
    	"metadata":  { "Latest user Lab Provision": null }
    }
}
```
- Expect to see the merged/updated values and not to see the deleted value.
